clients/include/testing_trtri_batched.hpp

```diff
@@ -36,33 +35,42 @@ void testing_trtri_batched(const Arguments& arg)
 
     // argument sanity check, quick return if input parameters are invalid before allocating invalid
     // memory
-    if(N < 0 || lda < 0 || lda < N || batch_count < 0)
+    if(N < 0 || lda < 0 || lda < N || batch_count <= 0)
```

Comment:
`batch_count == 0` is a valid size and not an error, just a no-op.

Reply:
We need `batch_count == 0` to work correctly. Usually this simply means returning `rocblas_status_success`, or `rocblas_status_size_unchanged` if `handle->is_device_memory_size_query()` is `true`.

I don't understand the code comments or your comments here ("we are checking for `rocblas_status_success` here when `batch_count ==0`"). You should not be _checking_ for `rocblas_status_success` when `batch_count == 0` -- you should be _returning_ `rocblas_status_success` (or `rocblas_status_size_unchanged`).

```
if(!batch_count)
    return handle->is_device_memory_size_query() ? rocblas_status_size_unchanged : rocblas_status_success;
```
Memory allocation and memory size queries are done in very specific ways. I can help.

It is not generally a good idea to ignore argument checking and go ahead and compute memory size based on possibly invalid arguments. Device memory size queries are allowed to return invalid argument errors.

The TRTRI _kernel_ code is not allowed to allocate device memory if it is called from another routine line TRSM. Its callers, like TRSM, must provide it buffer space. But a self-contained TRTRI which has a C wrapper, can allocate device memory. TRTRI is not a part of BLAS, so I don't know why we are extending a batch versioned for it. Only batched code to support batched TRSM is necessary. It's unfortunate that TRTRI was exposed in the public C interface, because there's no TRTRI in BLAS.

It took weeks to get allocation working correctly for TRTRI. It is very delicate code. As I recall, the batched code already existed when the device memory allocation code was added, but there were no C APIs for the batched versions.

I will need to look at this PR offline (not on GitHub) to properly review it. TRSM and TRTRI are the most complicated routines in BLAS.

Reply:
We are returning rocblas_status_success from trtri when batch_count == 0, in the testing code here we are _checking_ that that is the case. We aren't carrying out _all_ of the testing code when batch_count == 0 because this would complicate the code as the setup would have to have multiple checks that batch_count != 0 before allocating memory anyway, thus we are caling trtri early here and ensuring that it still returns rocblas_status_success. If we want to run through all of the testing code when batch_count == 0 we can do this, it might just add unnecessary complications to the test code and I'm unsure that it will be useful.

You mentioned that trtri kernel code is not allowed to allocate device memory. I am currently only allocating memory in the _impl functions, not in the _template as was recommended in the guidelines. So, when trsm calls trtri, for example, it will have to allocate memory before calling the _template. I'm not sure if you're referring to something else in another file that I'm missing.

The code for trtri_strided_batched had already existed, but was named trtri_batched, I changed the name of this to match the rest of the library. It was already present in the C api when I began this; I added trtri_strided_batched to the C api to be consistent.

Thanks for taking the time to review it offline, let me know if you have any further concerns that I can address.

Reply:
> We are returning rocblas_status_success from trtri when batch_count == 0, in the testing code here we are checking that that is the case. We aren't carrying out all of the testing code when batch_count == 0 because this would complicate the code as the setup would have to have multiple checks that batch_count != 0 before allocating memory anyway, thus we are caling trtri early here and ensuring that it still returns rocblas_status_success. If we want to run through all of the testing code when batch_count == 0 we can do this, it might just add unnecessary complications to the test code and I'm unsure that it will be useful.

Okay, if the testing code cannot currently handle `batch_count == 0` without a lot of guards, then maybe it's okay to do it in the same code as `batch_count < 0` or `n < 0` (with the "safe size" allocation of arrays), just with it expecting `rocblas_status_success` instead of `rocblas_status_invalid_size`.

> You mentioned that trtri kernel code is not allowed to allocate device memory. I am currently only allocating memory in the _impl functions, not in the _template as was recommended in the guidelines. So, when trsm calls trtri, for example, it will have to allocate memory before calling the _template. I'm not sure if you're referring to something else in another file that I'm missing.

That is the correct interpretation.

> The code for trtri_strided_batched had already existed, but was named trtri_batched, I  changed the name of this to match the rest of the library. It was already present in the C api when I began this; I added trtri_strided_batched to the C api to be consistent.

That is fine, to be consistent. I don't know how many users are actually calling the `TRTRI` C API. `TRTRI` is mainly important for `TRSM`, since it inverts the diagonal blocks.